sql-to-dbsp-compiler/SQL-compiler/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -0,0 +1,193 @@
+<#--
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to you under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+-->
+
+// This file is adapted from Calcite:
+// https://github.com/apache/calcite/tree/main/babel/src/main/codegen/includes/parserImpls.ftl
+
+JoinType LeftSemiJoin() :
+{
+}
+{
+    <LEFT> <SEMI> <JOIN> { return JoinType.LEFT_SEMI_JOIN; }
+}
+
+SqlNode DateaddFunctionCall() :
+{
+    final Span s;
+    final SqlOperator op;
+    final SqlIntervalQualifier unit;
+    final List<SqlNode> args;
+    SqlNode e;
+}
+{
+    (   <DATE_PART>  { op = SqlLibraryOperators.DATE_PART; }
+    |   <DATEADD> { op = SqlLibraryOperators.DATEADD; }
+    |   <DATEDIFF> { op = SqlLibraryOperators.DATEDIFF; }
+    |   <DATEPART>  { op = SqlLibraryOperators.DATEPART; }
+    )
+    { s = span(); }
+    <LPAREN> unit = TimeUnitOrName() {
+        args = startList(unit);
+    }
+    (
+        <COMMA> e = Expression(ExprContext.ACCEPT_SUB_QUERY) {
+            args.add(e);
+        }
+    )*
+    <RPAREN> {
+        return op.createCall(s.end(this), args);
+    }
+}
+
+boolean IfNotExistsOpt() :
+{
+}
+{
+    <IF> <NOT> <EXISTS> { return true; }
+|
+    { return false; }
+}
+
+TableCollectionType TableCollectionTypeOpt() :
+{
+}
+{
+    <MULTISET> { return TableCollectionType.MULTISET; }
```

Comment:
Do we need these?

Reply:
Not currently, but these may be useful. But see the comment above.